ui/src/components/Canvas.tsx

```diff
@@ -890,114 +878,86 @@ export function Canvas() {
-          ...currentNode,
-          // selected: false,
-          style: {
-            ...currentNode.style,
-            // boxShadow: `${userColor} 0px 15px 25px`,
-          },
-        });
-      }
-    },
-    [setNodes, userColor]
-  );
+  // FIXME: add awareness info when dragging
+  const onNodeDragStart = () => {};
 
   const onNodeDragStop = useCallback(
-    (event, node: Node) => {
+    // handle nodes list as multiple nodes can be dragged together at once
+    (event, _n: Node, nodes: Node[]) => {
       const reactFlowBounds = reactFlowWrapper.current.getBoundingClientRect();
       // This mouse position is absolute within the canvas.
       const mousePos = reactFlowInstance.project({
         x: event.clientX - reactFlowBounds.left,
         y: event.clientY - reactFlowBounds.top,
       });
+
+      const scope = getScopeAt(
+        mousePos.x,
+        mousePos.y,
+        nodes.map((n) => n.id)
+      );
+
+      if (!scope || scope.id === commonParent) return;
+
       // check if this position is inside parent scope
-      if (
-        mousePos.x < node.positionAbsolute!.x ||
-        mousePos.y < node.positionAbsolute!.y ||
-        mousePos.x > node.positionAbsolute!.x + node.width! ||
-        mousePos.y > node.positionAbsolute!.y + node.height!
-      ) {
-        // console.log("Cannot drop outside parent scope");
-        return;
+
+      // FIXME: a better way to do this: check if the commonParent is the ancestor of the scope
+
+      if (commonParent !== undefined && commonParent !== "ROOT") {
+        const currentParent = nodesMap.get(commonParent);
+        if (currentParent) {
+          if (
+            mousePos.x < currentParent.positionAbsolute!.x ||
+            mousePos.x >
+              currentParent.positionAbsolute!.x + currentParent.width! ||
+            mousePos.y < currentParent.positionAbsolute!.y ||
+            mousePos.y >
+              currentParent.positionAbsolute!.y + currentParent.height!
+          ) {
+            // the mouse is outside the current parent, the nodes can't be dragged out
+            // console.log("Cannot drop outside parent scope");
+            return;
+          }
+        }
       }
-      // Check which group is at this position.
-      const scope = getScopeAt(mousePos.x, mousePos.y, node.id);
-      let absX = node.position.x;
-      let absY = node.position.y;
-      if (scope) {
+
+      // check if this position is inside parent scope
+      nodes.forEach((node) => {
+        let absX = node.position.x;
+        let absY = node.position.y;
         console.log("dropped into scope:", scope);
+
         // compute the actual position
         let [dx, dy] = getAbsPos({ node: scope, nodesMap });
         absX = node.positionAbsolute!.x - dx;
         absY = node.positionAbsolute!.y - dy;
-      }
-      // first, dispatch this to the store
-      setPodPosition({
-        id: node.id,
-        x: absX,
-        y: absY,
-      });
```

Comment:
If we don't call `setPodPosition`, how would the store be updated? An immediate issue I observe: if you move a pod, the dirty status is not set, and thus the cloud sync is not triggered. However, when I refreshed, the new position indeed persisted. I guess this persistence is from the yjs side, is it? If so, does it mean that there will be some inconsistency between yjs server and database?

Reply:
It is indeed inconsistent between yjs server and the database. If you restart the yjs socket container, you can see all the positions are wrong. The positions are not saved to the database.

Reply:
Yes, it only updates `nodesMap`, notice that `nodesMap` is shared by all online users via the yjs server, they pull the update from `nodesMap` and set nodes in canvas according to it immediately. When `nodesMap` is connected during initialization, it first syncs with `nodesMap`, it only pulls data from DB when no data in  `nodesMap`. In the current stage, the position changes are not written to the DB.

I agree with your concern. We should recover those lines. That is actually what I mentioned in known issue 3, in the past, we don't do any position update in `store.pods[id]` when the dragging is from other users. Fortunately, those lines toggles DB update on one user-side. But other users should also be subscribed to the change and update their `store.pods`, or the data inconsistency issue become more terrible.

Reply:
> I agree with your concern. We should recover those lines

I've recovered the `setPodPosition` call in #137.

> a user may drag a node into a scope and update everything as expected on their own side, but the collaborator users can only view the change on front-end side, store.pods on their side doesn't update.

If I understand correctly, when a user drags a node into a scope:
- `yjs.nodesMap` will be updated (thus the graph itself is updated)
- `store.pods[id].parent` (and `store.pods[id].children`) won't be updated

If so, we need to fix it, otherwise, the scoped runtime won't work correctly. I think your proposed useEffect method is the way to go. If you have the bandwidth, please go ahead. Otherwise, I can do it as well.

> in the past, we don't do any position update in store.pods[id] when the dragging is from other users

I feel this is OK. We don't need to update store.pods[id].position, because it is not used to render the nodes. On refresh, it will be loaded from DB correctly.

Reply:
> If I understand correctly, when a user drags a node into a scope:
> - yjs.nodesMap will be updated (thus the graph itself is updated)
> - store.pods[id].parent (and store.pods[id].children) won't be updated
> If so, we need to fix it, otherwise, the scoped runtime won't work correctly. I think your proposed useEffect method is > the way to go. If you have the bandwidth, please go ahead. Otherwise, I can do it as well.

Yes, you got it exactly. Thanks a lot.

> I feel this is OK. We don't need to update store.pods[id].position, because it is not used to render the nodes. On refresh, it will be loaded from DB correctly.

Imagine a situation:

1. a pod is dragged from a place to another place by a user other than myself, `store.pod[id].position` is not updated on my side, which is `(oldX, oldY)`, but it should be `(oldX, oldY)`
2. later, I change something inside this pod, e.g., name, code, calling `setPodName` or `setPodContent`, which updates only name or content, keeps position as  `(oldX, oldY)`, and mark the pod as dirty.
3. `remoteUpdateAllPods` is called periodically to write this dirty pod to the database, which is `{name: newName, content: newContent,  position: oldX, oldY}`

That's what I mean by saying data inconsistency issue.

Besides, I'm now considering copy-paste multi-selected pods, I may use serialized `store.pod[id]` info in clipboard, their relative positions may also be kept after pasting.

Reply:
So `store.pods[id].content` should also have this same inconsistency issue (as in your example step 1,2,3), right?

Reply:
> So `store.pods[id].content` should also have this same inconsistency issue (as in your example step 1,2,3), right?

I think the code itself is okay. Because we use `setPodContent` in editor's `onChange`, it should be called every time the value updates, even if the update come from the yjs server. Of course, it may cause redundant writing to the DB, but not so significant in the current stage. 

I'm mainly worrying about the lazy update in position and parent info.

Reply:
> Because we use `setPodContent` in editor's `onChange`, it should be called every time the value updates, even if the update comes from the yjs server.

I see. Yeah, essentially we implement the useEffect update on `pods.content`. We should do the same for `position` and `parent` as well.

> Of course, it may cause redundant writing to the DB, but the performance overhead in it is not so critical in the current stage.

Agree, no need to worry about this.

> _BTW, I come up with a possible solution to mitigate the redundant DB update overhead caused by `nodesMap` changes: add a field in node's data, as `lastModifier` to track who pushed the update

Great idea; let's implement it later in this way.